### ISSUE TITLE
CCrypto::SymmetricEncrypt() findings

### ISSUE THREAD

**Issue by catid** [ValveSoftware/GameNetworkingSockets#16]
!!! This appears to be implementing AES-CBC, which is frowned upon:
http://latacora.singles/2018/04/03/cryptographic-right-answers.html
> Avoid: AES-CBC

Pad length round up to the next multiple of 16
!!! When even multiple this will pad an extra 16 unnecessarily

!!! These pointer comparisons are not supported by C++.  Some data may get corrupted in some builds of the software:
if ( ( pubPlaintextData > pubEncryptedData && pubPlaintextData < pubEncryptedData + cubEncryptedData ) ||
		 ( pubEncryptedData > pubPlaintextData && pubEncryptedData < pubPlaintextData + cubPlaintextData ) )
https://stackoverflow.com/questions/9086372/how-to-compare-pointers#9086675
> If two pointers p and q of the same type point to different objects that are not members of the same object or elements of the same array or to different functions, or if only one of them is null, the results of p<q, p>q, p<=q, and  p>=q are unspecified.

BExtractAESRoundKeys( &key, false, roundKeysAsU32, &nRounds )
!!! This function assumes 256 bit keys: AES_set_encrypt_key( rgchTest, 256, &keyTest );
But the AES_set_encrypt_key() function above uses cubKey * 8.
I'm worried that when cubKey is not 32, this will actually use an empty (zero) key or fail.

**Comment by fletcherdvalve** [comment]
There's some really good feedback in here, thanks for this.

I'm going to hold of responding or fixing because I'm working on rewriting the reliability layer, and some of this is going to change.  (E.g. moving the sequence number backwards and dealing with duplicate sequence numbers and replay packets.)

My initial thought on replay attacks:  I have thought about this in the past, and I was not convinced that the cost of protecting against replay attacks is clearly better than allowing duplicate packets to be processed further.  Especially if some general purpose across-the-boards rate limiting is employed, which I believe is necessary.  (The relays provide this protection for relayed connections -- that isn't part of the opensource code --- but I just realized that I need to bring over a similar protection for standard UDPv4 connections.)  Anybody who can observe the traffic likely has many other ways that they can disrupt that specific connection.  I don't know if I want to include that in the threat model we are trying to protect against.  I do believe that we should try to protect against is somebody getting connected and then trying to DoS the peer, nuking their connection and all other connections at the same time.  That is definitely in the threat model.  However, the ability to disrupt a connection that you are already able to observe -- I am not sure that is worthwhile to try to prevent that.

I'd be interested in your thoughts on that.

But I am changing the way the end-to-end sequence numbers are processed by the stats collection class, and the SNP layer, so I don't want to fix any of that until I have finished that big rewrite.

We might make some changes to the encryption mode and padding.  I think the "extra" padding is in there for a reason.  Something about a side channel timing attack.  But that reason may not be applicable in this code.  I'll keep this issue open and we might revisit that.  A wasted 16 bytes on the wire is definitely worth figuring out.

**Comment by catid** [comment]
For games being able to replay packets seems bad for example on Gunbound I was able to take multiple turns in a row by replaying the it’s your turn packet

**Comment by catid** [comment]
Agreed - It's totally valid to say that a feature does not belong at the transport layer and should be done up the stack.  "Handled by documentation."

**Comment by vlv-henryg** [comment]
I'm going to weigh in on the padding notes - CBC mode requires padding, but also an indicator of how long the padding is. Traditional padding schemes always add 1 to 16 bytes, and the value of the final byte tells you how much padding there is. As for the padding checks, we are permissive with the padding because it is explicitly not supposed to be an integrity check. That's the job of an HMAC or other authenticator. (Because of peculiar details of CBC mode, being very strict about padding is a good way to accidentally create a CBC padding oracle that can be used to decrypt arbitrary blocks by reodering them to be the last block, XORing specific bits in the previous block, and finding out if the last block was interpreted as valid padding or not.)

I won't say that CBC is a great choice, because it's not; if we were designing from scratch, I think ideally using a vetted high-level library like libsodium would be the way to go. If we had to stick to low-level primitives, AES-GCM would be better since it authenticates simultaneously, instead of being a separate step. But at least I don't think that CCrypto is doing a bad job at AES-CBC.

**Comment by fletcherdvalve** [comment]
I don't want to make out-of-order guarantees, because I think only the application knows whether that out-of-order data is useful or not.  (Or maybe add some sort of configuration option to block them.)  But duplicate messages -- there's probably universally no value to those, and it's pretty cheap to detect, and it eliminates a whole class of bugs and attacks.  So you're probably right that they should just be prevented.

**Comment by fletcherdvalve** [comment]
I've just checked in rewrite of the reliability layer.  It is a lot more robust against packet number shenanigans.  It checks for failed decryption earlier, and won't ever decrement the packet number -- that was only necessary before in case the wrapping numbers got way off track, but now I'm handling that differently.

I actually did not yet specifically guard against duplicate packets, though...  I'll open up another issue to remember to do that for that and close this one.

Thanks again for your feedback.
